### PR TITLE
Metadata reader

### PR DESCRIPTION
**Pull request by DanielElisenberg**
# METADATA READER

Now that we inject centrally defined attributes and unit type variables, the validate_metadata function must also produce a product to working_dir. This changes the API of the package.

### What's new?
* Reworked dataset_reader into metadata_reader + dataset_reader
* The basic validate function works the same way but the code is updated to use both steps
* The validate_metadata function now takes in more arguments and acts more similarly to validate
* Moved some directory generation and cleanup logic to local_storage
* Updated tests and docs

### CODE REVIEW

File: tests/unit/test_validate_metadata.py [hunk 1]
```diff
@@ -8,54 +11,80 @@
...
+            encoding='utf-8'
+        ) as f:
+            actual_metadata = json.load(f)
+        with open(
+            Path(EXPECTED_DIRECTORY) / f'{metadata}.json',
+            'r',
+            encoding='utf-8'
+        ) as f:
+            expected_metadata = json.load(f)
+        assert actual_metadata == expected_metadata
```
Comment (pawbu) [inline tests/unit/test_validate_metadata.py:48]: Unless I'm missing something: two files (actual_metadata and expected_metadata) are read and compared, no custom logic here that is being tested 🤔 The other assert is OK though.

File: tests/unit/test_validate_metadata.py [hunk 2]
```diff
@@ -8,54 +11,80 @@
...
+        encoding='utf-8'
+    ) as f:
+        actual_metadata = json.load(f)
+    with open(
+        Path(EXPECTED_DIRECTORY) / f'{VALID_METADATA_REF}.json',
+        'r',
+        encoding='utf-8'
+    ) as f:
+        expected_metadata = json.load(f)
+    assert actual_metadata == expected_metadata
```
Comment (pawbu) [inline tests/unit/test_validate_metadata.py:80]: Same here?

**Review by pawbu — APPROVED** [review]
Looks good, one question

File: tests/unit/test_validate_metadata.py (hunk 1 above)
Comment (DanielElisenberg) [reply tests/unit/test_validate_metadata.py:48]: The actual_metadata is the metadata that is produced by the validate_metadata function, while the expected_metadata is what the validate_metadata is expected to produce based on the input data. So its similar to:

```py
expected_amount = 8
actual_amount = count_people_in_room(ROOM_WITH_8_PEOPLE)
assert actual_amount == expected_amount
```

Comment (pawbu) [reply tests/unit/test_validate_metadata.py:48]: Right, validate_metadata creates a file that you read in the test 👍

Comment (DanielElisenberg) [reply tests/unit/test_validate_metadata.py:48]: That being said: We might need to update these tests in the future, they might be too statically defined and stringent. If we update a descriptive field in the central unit type definitions they will fail f.ex. Let's keep an eye on them 👀

**Comment by DanielElisenberg** [comment]
Reverted back to 5.0.0 to delay release of 6.0.0 until the addition of a error_threshold

**Pull request merged**